proof/refine/ARM/Ipc_R.thy

```diff
@@ -4604,68 +4493,446 @@ lemma maybe_return_sc_weak_valid_sched_action:
+   apply (frule (1) idle_no_ex_cap)
+   apply (clarsimp simp: valid_idle_def)
+  apply (simp add: receive_ipc_def receiveIPC_def)
+  apply add_sym_refs
+  apply (case_tac cap, simp_all add: isEndpointCap_def)
+  apply (rename_tac word1 word2 right)
+  apply (rule corres_stateAssert_assume)
+   apply (rule stronger_corres_guard_imp)
+   apply (subst receiveIPC_corres_helper)
+     apply (clarsimp simp: receive_ipc_preamble_def[symmetric] receiveIPC_preamble_def[symmetric])
+     apply (rule corres_split_deprecated[OF _ receiveIPC_preamble_corres])
+         apply (rule corres_split_deprecated [OF _ get_ep_corres])
+           apply (rename_tac ep ep')
+           apply (rule corres_split_deprecated [OF _ gbn_corres])
+             apply (rule_tac r'="ntfn_relation" in corres_split_deprecated)
+                apply (rule corres_if)
+                  apply (clarsimp simp: ntfn_relation_def Ipc_A.isActive_def Endpoint_H.isActive_def
+                                 split: Structures_A.ntfn.splits Structures_H.notification.splits)
+                 apply (simp only: )
+                 apply (rule complete_signal_corres)
+                apply (rule corres_split_deprecated[where r'=dc])
+                   apply (rule_tac P="einvs and ?vs and ?tat and ?slot3 and ?tex
+                                            and ep_at word1 and valid_ep ep and ko_at (Endpoint ep) word1
+                                            and receive_ipc_preamble_rv reply_cap replyOpt
+                                            and ?rrefs" and
+                                   P'="invs' and tcb_at' thread and ep_at' word1
+                                             and valid_ep' ep'
+                                             and valid_bound_reply' replyOpt"
+                                   in corres_inst)
+                   apply (rule_tac P'="valid_bound_obj' valid_replies'_sc_asrt replyOpt" and
+                                   P=\<top> in corres_add_guard)
+                    apply (case_tac replyOpt; simp)
+                    apply (erule valid_replies_sc_cross; clarsimp elim!: reply_at_ppred_reply_at)
+                   apply (case_tac ep)
+                     \<comment> \<open>IdleEP\<close>
+                     apply (simp add: ep_relation_def)
+                     apply (fold dc_def)[1]
+                     apply (rule corres_guard_imp)
+                       apply (case_tac isBlocking; simp)
+                        apply (rule corres_split_deprecated [OF _ sts_corres])
+                           apply (rule corres_split_deprecated[where r'=dc])
+                              apply (rule set_ep_corres)
+                              apply (simp add: ep_relation_def)
+                             apply (rule corres_when, clarsimp)
+                             apply (rule replyTCB_update_corres)
+                            prefer 6 \<comment> \<open> defer wp until corres complete \<close>
```

Comment:
Maybe you can use an Eisbach `match` statement to pick out the `corres` statement here instead of the `prefer`?

Reply:
I tried to do this, but maybe there's a problem associated iwth schematics? I couldn't make it work.

Reply:
It's hard to tell without looking at the state, but using `corres_split` instead of `corres_split_deprecated` might let you avoid needing this `prefer`. One of the reasons that I made that change was because I found it led to a more natural order of subgoals.